File: ios/Objective-C/Usecase/TinyBarcode/TinyBarcode/Views/CameraSettingView.m
#import "CameraSettingView.h"

@interface CameraSettingView ()

@property (nonatomic, strong) UILabel *autoZoomLabel;

@property (nonatomic, strong) UISwitch *controlSwitch;

@end

@implementation CameraSettingView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        [self setupUI];
    }
    return self;
}

- (void)setupUI {
    self.backgroundColor = [UIColor colorWithRed:40/255.0 green:40/255.0 blue:40/255.0 alpha:1];
    
    CGFloat contentHeight = KCameraSettingAvailableHeight;
    
    self.autoZoomLabel = [[UILabel alloc] init];
    self.autoZoomLabel.frame = CGRectMake(kLeftMarginOfContainer, 0, 100, contentHeight);
    self.autoZoomLabel.text = @"Auto Zoom";
    self.autoZoomLabel.font = kFont_SystemDefault(KCameraSettingTitleTextSize);
    self.autoZoomLabel.textColor = [UIColor whiteColor];
    self.autoZoomLabel.textAlignment = NSTextAlignmentLeft;
    [self addSubview:self.autoZoomLabel];
    
    self.controlSwitch = [[UISwitch alloc] init];
    self.controlSwitch.left = kScreenWidth - kRightMarginOfContainer - self.controlSwitch.width;
    self.controlSwitch.top = (contentHeight - self.controlSwitch.height) / 2.0;
    self.controlSwitch.onTintColor = kSwitchOnTintColor;
    self.controlSwitch.tintColor = kSwitchOffTintColor;
    self.controlSwitch.backgroundColor = kSwitchOffTintColor;
    self.controlSwitch.layer.cornerRadius = self.controlSwitch.height / 2.0;
    [self.controlSwitch addTarget:self action:@selector(controlSwitchChanged:) forControlEvents:UIControlEventValueChanged];
    [self addSubview:self.controlSwitch];
}

- (void)controlSwitchChanged:(UISwitch *)controlSwitch
{
    if (controlSwitch.on) {
        self.controlSwitch.thumbTintColor = kSwitchOnThumbColor;
    } else {
        self.controlSwitch.thumbTintColor = kSwitchOffThumbColor;
    }
    if (self.switchChangedBlock) {
        self.switchChangedBlock(controlSwitch.on);
    }
}

- (void)updateSwitchState:(BOOL)switchState {
    self.controlSwitch.on = switchState;
    
    if (self.controlSwitch.on) {
        self.controlSwitch.thumbTintColor = kSwitchOnThumbColor;
    } else {
        self.controlSwitch.thumbTintColor = kSwitchOffThumbColor;
    }
}

@end

File: ios/UseCase/TinyBarcodeDecoding/TinyBarcodeDecoding/Views/CameraSettingView.swift
//
//  CameraSettingView.swift
//  TinyBarcode
//
//  Created by dynamsoft's mac on 2022/11/25.
//

import UIKit

typealias SwitchChangedCompletion = (_ isOn: Bool) -> Void

class CameraSettingView: UIView {

    var switchChangedCompletion: SwitchChangedCompletion?
    
    private var autoZoomLabel: UILabel = {
        let label = UILabel.init(frame: CGRect(x: kLeftMarginOfContainer, y: 0, width: 100, height: KCameraSettingViewAvailableHeight))
        label.textColor = .white
        label.text = "Auto Zoom"
        label.font = kFont_Regular(KCameraSettingTitleTextSize)
        label.textAlignment = .left
        return label
    }()
    
    private var controlSwitch: UISwitch = {
        let controlSwitch = UISwitch.init()
        controlSwitch.left = kScreenWidth - kRightMarginOfContainer -  controlSwitch.width
        controlSwitch.top = (KCameraSettingViewAvailableHeight - controlSwitch.height) / 2.0
        controlSwitch.onTintColor = kSwitchOnTintColor
        controlSwitch.tintColor = kSwitchOffTintColor
        controlSwitch.backgroundColor = kSwitchOffTintColor
        controlSwitch.layer.cornerRadius = controlSwitch.height / 2.0
        return controlSwitch
    }()
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupUI()
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
    }
    
    private func setupUI() -> Void {
        self.backgroundColor = UIColor(red: 40 / 255.0, green: 40 / 255.0, blue: 40 / 255.0, alpha: 1)
        self.addSubview(autoZoomLabel)
        self.addSubview(controlSwitch)
        controlSwitch.addTarget(self, action: #selector(controlSwitchChange(_:)), for: .valueChanged)
    }
    
    func updateSwitch(with switchState: Bool) -> Void {
        self.controlSwitch.isOn = switchState
        if (controlSwitch.isOn) {
            controlSwitch.thumbTintColor = kSwitchOnThumbColor
        } else {
            controlSwitch.thumbTintColor = kSwitchOffThumbColor
        }
    }
    
    @objc private func controlSwitchChange(_ controlSwitch: UISwitch) -> Void {
        if (controlSwitch.isOn) {
            controlSwitch.thumbTintColor = kSwitchOnThumbColor
        } else {
            controlSwitch.thumbTintColor = kSwitchOffThumbColor
        }
        
        self.switchChangedCompletion?(controlSwitch.isOn)
    }

}
